app/src/main/java/de/westnordost/streetcomplete/quests/smoking/AddSmoking.kt

```diff
@@ -0,0 +1,77 @@
+package de.westnordost.streetcomplete.quests.smoking
+
+import de.westnordost.streetcomplete.R
+import de.westnordost.streetcomplete.data.meta.isShopExpressionFragment
+import de.westnordost.streetcomplete.data.meta.updateWithCheckDate
+import de.westnordost.streetcomplete.data.osm.mapdata.Element
+import de.westnordost.streetcomplete.data.osm.mapdata.MapDataWithGeometry
+import de.westnordost.streetcomplete.data.osm.mapdata.filter
+import de.westnordost.streetcomplete.data.osm.osmquests.OsmFilterQuestType
+import de.westnordost.streetcomplete.data.osm.osmquests.Tags
+import de.westnordost.streetcomplete.data.user.achievements.QuestTypeAchievement.CITIZEN
+import de.westnordost.streetcomplete.ktx.arrayOfNotNull
+import de.westnordost.streetcomplete.ktx.containsAny
+
+class AddSmoking : OsmFilterQuestType<SmokingAllowed>() {
+
+    private val elementFilterBasic = """
+        (
+            amenity ~ bar|cafe|pub|biergarten|restaurant|food_court|nightclub|stripclub
+            or leisure ~ outdoor_seating
+            or (
+                (amenity ~ fast_food|ice_cream or shop ~ ice_cream|deli|bakery|coffee|tea|wine)
+                and (
+                    (outdoor_seating and outdoor_seating != no)
+                    or (indoor_seating and indoor_seating != no)
+                )
+            )
+        )
+    """
+
+    override val elementFilter = """
+         nodes, ways, relations with
+         ${elementFilterBasic}
+         and takeaway != only
+         and (!smoking or smoking older today -8 years)
+    """
+
+    override val changesetComment = "Add smoking status"
+    override val wikiLink = "Key:smoking"
+    override val icon = R.drawable.ic_quest_smoke
+    override val isReplaceShopEnabled = true
+    override val defaultDisabledMessage = R.string.default_disabled_msg_go_inside_regional_warning
+
+    override val questTypeAchievements = listOf(CITIZEN)
+
+    override fun getTitle(tags: Map<String, String>) =
+        if (hasProperName(tags))
+            R.string.quest_smoking_name_type_title
+        else
+            R.string.quest_smoking_no_name_title
+
+    override fun getTitleArgs(tags: Map<String, String>, featureName: Lazy<String?>): Array<String> {
+        val name = tags["name"] ?: tags["brand"] ?: tags["operator"]
+        return arrayOfNotNull(name, featureName.value)
+    }
+
+    /* note: copied from IS_SHOP_OR_DISUSED_SHOP_EXPRESSION as we seemingly can't extend it with extra strings */
+    override fun getHighlightedElements(element: Element, getMapData: () -> MapDataWithGeometry) =
+        getMapData().filter("""
+            nodes, ways, relations with
+              ${isShopExpressionFragment()}
+              or ${isShopExpressionFragment("disused")}
+              or shop = vacant
+              or ${elementFilterBasic}
+        """)
```

Comment:
I do not think this is worth duplicating the code here. What cases are caught with this over `IS_SHOP_OR_DISUSED_SHOP_EXPRESSION`? 

As far as I see, just `amenity ~ food_court|nightclub|stripclub or leisure = outdoor_seating`. I wouldn't expect any of these to be right next to each other. Maybe two nightclubs, but this is a really rare situation.

Actually, I wonder if `nightclub` and `stripclub` could not be added to "isKindOfShop" filter instead. Aren't those typically in the same kind of rented properties as pubs, bars etc. would be?

Reply:
> As far as I see, just amenity ~ food_court|nightclub|stripclub or leisure = outdoor_seating. I wouldn't expect any of these to be right next to each other. Maybe two nightclubs, but this is a really rare situation.

Well, `leisure=outdoor_seating` especially are quite often near one another, as they are commonly in front of cafes, which are often (at least down here) crowded together esp. in downtowns, eg. [overpass](http://overpass-turbo.eu/?Q=%2F*%0AThis%20has%20been%20generated%20by%20the%20overpass-turbo%20wizard.%0AThe%20original%20search%20was%3A%0A%E2%80%9Cleisure%20%3D%20outdoor_seating%E2%80%9D%0A*%2F%0A%5Bout%3Ajson%5D%5Btimeout%3A25%5D%3B%0A%2F%2F%20gather%20results%0A(%0A%20%20%2F%2F%20query%20part%20for%3A%20%E2%80%9Cleisure%3Doutdoor_seating%E2%80%9D%0A%20%20nwr%5B%22leisure%22%3D%22outdoor_seating%22%5D(%7B%7Bbbox%7D%7D)%3B%0A%0A)%3B%0A%2F%2F%20print%20results%0Aout%20body%3B%0A%3E%3B%0Aout%20skel%20qt%3B&C=45.81196;15.97645;17&R)

So I'd like to **_at least_** _additionally_ display  `leisure=outdoor_seating`. `nightclub` and `stripclub` could  be added to `isKindOfShop` IMO (but I'm not an expert on all places where it is used in code) - but that would not solve my problem by itself, as I want to add at least that one tag more.

Ideas I've pondered (besides current copy/paste implementation which I'm not terribly happy about either):
- just filter on `elementFilterBasic`  (i.e. only highlight places which could be asked by this quest)
- extend  `isKindOfShop` - as you said, it is OK for `nightclub` and `stripclub`, but how about `leisure=outdoor_seating`? Anyway I would not touch code outside `smoking/` directory without your approval
- create `IS_SHOP_OR_DISUSED_SHOP_EXPRESSION` class/object (or how is it called in kotlin) from  `IS_SHOP_OR_DISUSED_SHOP_EXPRESSION_STRING` instead; and then I could use simple `${IS_SHOP_OR_DISUSED_SHOP_EXPRESSION_STRING} or ${elementFilterBasic}` (or is there some other way to merge them in kotlin?). Then, this and other quests could extend that default with just a thing or two they need, without this ugly copy/paste. But again, it is outside my code, so I'd need to know if you'd be OK with that change. i.e.
```diff
- val IS_SHOP_OR_DISUSED_SHOP_EXPRESSION = """
+ val IS_SHOP_OR_DISUSED_SHOP_EXPRESSION_STR = """
    nodes, ways, relations with
      ${isShopExpressionFragment()}
      or ${isShopExpressionFragment("disused")}
      or shop = vacant
- """.toElementFilterExpression()
+ """
+ val IS_SHOP_OR_DISUSED_SHOP_EXPRESSION = IS_SHOP_OR_DISUSED_SHOP_EXPRESSION_STR.toElementFilterExpression()
```
- Or maybe just `.toElementFilterExpression()` could be dropped from existing `IS_SHOP_OR_DISUSED_SHOP_EXPRESSION` ? Majority of uses seem to be `getMapData().filter` which as I understand can also take plain string as an argument, so they can continue to use it unchanged; and few other places could add `.toElementFilterExpression()` themselves if they need it (excuse my naivete if that is impossible/undesirable)

Reply:
Hm ok, I'd find best if:
- strip and night club is added to the "shop expression"
- leave the code in `AddSmoking` as is, mainly because of `leisure=outdoor_seating`

Reply:
OK, added in PR https://github.com/streetcomplete/StreetComplete/pull/3822
Leaving `smoking/` code as-is.

Reply:
Or actually... please just 
- add strip and night club to the "shop expression"
- use IS_SHOP_OR_DISUSED_SHOP_EXPRESSION only in getHighlightedElements

Reason: highlighting outdoor seatings may actually confuse the user, as there seem to be unnamed shops close by in e.g. the example you posted. The user will be used to seeing shops etc. highlighted for these kinds of quests

Reply:
as you wish. done.